Flaky test: GrizzlyLDAPListenerTestCase#testLDAPListenerLoadBalanceDuringHandleBind times out on CI

### Summary

`GrizzlyLDAPListenerTestCase#testLDAPListenerLoadBalanceDuringHandleBind` (module `opendj-grizzly`) intermittently fails on CI with a `TimeoutException`, breaking builds of unrelated PRs.

### Example failure

Run: https://github.com/OpenIdentityPlatform/OpenDJ/actions/runs/30247170643/job/89916711062 (`build-maven (ubuntu-latest, 26)`, PR #761)

```
[ERROR] org.forgerock.opendj.grizzly.GrizzlyLDAPListenerTestCase.testLDAPListenerLoadBalanceDuringHandleBind -- Time elapsed: 10.02 s <<< FAILURE!
java.util.concurrent.TimeoutException
	at org.forgerock.util.promise.PromiseImpl.await(PromiseImpl.java:636)
	at org.forgerock.util.promise.PromiseImpl.get(PromiseImpl.java:138)
	at org.forgerock.opendj.grizzly.GrizzlyLDAPListenerTestCase.testLDAPListenerLoadBalanceDuringHandleBind(GrizzlyLDAPListenerTestCase.java:433)
```

The only failure out of 1038 tests in the module; the same run passed on `macos-latest` and `windows-latest` with the same JDK, and the triggering PR (#761) does not touch `opendj-grizzly` at all — so this is environmental, not a regression.

### Analysis

The test starts a load-balanced proxy in front of an intentionally-offline server and an online server, performs a bind through the proxy, and then waits for the connection contexts:

```java
assertThat(proxyServerConnection.context.get(10, TimeUnit.SECONDS)).isNotNull();
assertThat(onlineServerConnection.context.get(10, TimeUnit.SECONDS)).isNotNull();   // line 433
```

On slow GitHub-hosted runners (4 CPUs, 512 MB test-JVM heap) the failover from the offline server to the online one occasionally does not complete within the 10-second window, so the `context.get(10, TimeUnit.SECONDS)` at line 433 times out.

### Suggested fix

Either of:
- increase the wait window for these asserts (e.g. 60 s — the value only matters on genuine failure), or
- add a retry analyzer / `@Test(retryAnalyzer = …)` for this timing-sensitive test.

--- a/opendj-grizzly/src/test/java/org/forgerock/opendj/grizzly/GrizzlyLDAPListenerTestCase.java
+++ b/opendj-grizzly/src/test/java/org/forgerock/opendj/grizzly/GrizzlyLDAPListenerTestCase.java
@@ -28,7 +28,9 @@ import static org.forgerock.opendj.ldap.TestCaseUtils.*;
 import static org.forgerock.util.Options.defaultOptions;
 import static org.mockito.Mockito.mock;
 
+import java.io.IOException;
 import java.net.InetSocketAddress;
+import java.net.ServerSocket;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.concurrent.CountDownLatch;
@@ -187,6 +189,18 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
         }
     }
 
+    /**
+     * Binds a server socket on a free loopback port and keeps it bound so that the port cannot be
+     * handed out again to a listener which is created later. The socket must be closed before
+     * connection attempts which expect a connection failure are performed against its address.
+     */
+    private static ServerSocket reserveSocketAddress() throws IOException {
+        final ServerSocket socket = new ServerSocket();
+        socket.setReuseAddress(true);
+        socket.bind(loopbackWithDynamicPort());
+        return socket;
+    }
+
     /** Disables logging before the tests. */
     @BeforeClass
     public void disableLogging() {
@@ -285,14 +299,19 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
                         onlineServerConnectionFactory));
         final InetSocketAddress onlineAddr = onlineServerListener.firstSocketAddress();
 
+        // Reserve the offline server ports, keeping them bound until the proxy listener is
+        // created: a port released too early could be handed out to the proxy listener itself,
+        // making the load balancer connect back to the proxy instead of the online server.
+        final ServerSocket offlineServerSocket1 = reserveSocketAddress();
+        final ServerSocket offlineServerSocket2 = reserveSocketAddress();
         try {
             // Connection pool and load balancing tests.
-            InetSocketAddress offlineAddress1 = findFreeSocketAddress();
+            InetSocketAddress offlineAddress1 = (InetSocketAddress) offlineServerSocket1.getLocalSocketAddress();
             final ConnectionFactory offlineServer1 =
                     Connections.newNamedConnectionFactory(new LDAPConnectionFactory(
                             offlineAddress1.getHostName(),
                             offlineAddress1.getPort()), "offline1");
-            InetSocketAddress offlineAddress2 = findFreeSocketAddress();
+            InetSocketAddress offlineAddress2 = (InetSocketAddress) offlineServerSocket2.getLocalSocketAddress();
             final ConnectionFactory offlineServer2 =
                     Connections.newNamedConnectionFactory(new LDAPConnectionFactory(
                             offlineAddress2.getHostName(),
@@ -329,11 +348,15 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
                     new ServerConnectionFactoryAdapter(Options.defaultOptions().get(LDAP_DECODE_OPTIONS),
                             proxyServerConnectionFactory));
             final InetSocketAddress proxyAddr = proxyListener.firstSocketAddress();
+            // Release the reserved ports now that every listener is bound: connection attempts
+            // to the offline addresses must fail with a connection error.
+            offlineServerSocket1.close();
+            offlineServerSocket2.close();
+            final LDAPConnectionFactory proxyClientFactory =
+                    new LDAPConnectionFactory(proxyAddr.getHostName(), proxyAddr.getPort());
             try {
                 // Connect and close.
-                final Connection connection =
-                        new LDAPConnectionFactory(proxyAddr.getHostName(),
-                            proxyAddr.getPort()).getConnection();
+                final Connection connection = proxyClientFactory.getConnection();
 
                 assertThat(proxyServerConnection.context.get(10, TimeUnit.SECONDS)).isNotNull();
                 assertThat(onlineServerConnection.context.get(10, TimeUnit.SECONDS)).isNotNull();
@@ -343,9 +366,13 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
 
                 proxyServerConnection.isClosed.await();
             } finally {
+                proxyClientFactory.close();
+                loadBalancer.close();
                 proxyListener.close();
             }
         } finally {
+            offlineServerSocket1.close();
+            offlineServerSocket2.close();
             onlineServerListener.close();
         }
     }
@@ -368,14 +395,19 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
                         onlineServerConnectionFactory));
         final InetSocketAddress onlineServerAddr = onlineServerListener.firstSocketAddress();
 
+        // Reserve the offline server ports, keeping them bound until the proxy listener is
+        // created: a port released too early could be handed out to the proxy listener itself,
+        // making the load balancer connect back to the proxy instead of the online server.
+        final ServerSocket offlineServerSocket1 = reserveSocketAddress();
+        final ServerSocket offlineServerSocket2 = reserveSocketAddress();
         try {
             // Connection pool and load balancing tests.
-            InetSocketAddress offlineAddress1 = findFreeSocketAddress();
+            InetSocketAddress offlineAddress1 = (InetSocketAddress) offlineServerSocket1.getLocalSocketAddress();
             final ConnectionFactory offlineServer1 =
                     Connections.newNamedConnectionFactory(
                             new LDAPConnectionFactory(offlineAddress1.getHostName(),
                                     offlineAddress1.getPort()), "offline1");
-            InetSocketAddress offlineAddress2 = findFreeSocketAddress();
+            InetSocketAddress offlineAddress2 = (InetSocketAddress) offlineServerSocket2.getLocalSocketAddress();
             final ConnectionFactory offlineServer2 =
                     Connections.newNamedConnectionFactory(
                             new LDAPConnectionFactory(offlineAddress2.getHostName(),
@@ -421,11 +453,15 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
                             proxyServerConnectionFactory));
             final InetSocketAddress proxyAddr = (InetSocketAddress) proxyListener.firstSocketAddress();
 
+            // Release the reserved ports now that every listener is bound: connection attempts
+            // to the offline addresses must fail with a connection error.
+            offlineServerSocket1.close();
+            offlineServerSocket2.close();
+            final LDAPConnectionFactory proxyClientFactory =
+                    new LDAPConnectionFactory(proxyAddr.getHostName(), proxyAddr.getPort());
             try {
                 // Connect, bind, and close.
-                final Connection connection =
-                        new LDAPConnectionFactory(proxyAddr.getHostName(),
-                            proxyAddr.getPort()).getConnection();
+                final Connection connection = proxyClientFactory.getConnection();
                 try {
                     connection.bind("cn=test", "password".toCharArray());
 
@@ -439,9 +475,13 @@ public class GrizzlyLDAPListenerTestCase extends SdkTestCase {
                 // Wait for connect/close to complete.
                 proxyServerConnection.isClosed.await();
             } finally {
+                proxyClientFactory.close();
+                loadBalancer.close();
                 proxyListener.close();
             }
         } finally {
+            offlineServerSocket1.close();
+            offlineServerSocket2.close();
             onlineServerListener.close();
         }
     }

